### PR TITLE
Assignment4

### PR DESCRIPTION
**Pull request by nfpaiva**

### CODE REVIEW

**Review by andrefnb — APPROVED** [review]
Parece-me tudo bem! Não encontro problemas nenhuns com tyhints ou paths e coisas do género. Até incluis-te imensos testes de falta de ficheiros etc. E fizeste-me aperceber que me falta umas coisinhas. Fizeste também classes de contexto tipo FileHandler que abstraem as chamadas ao strategy pattern. Overall it is very good! Congratulations!

File: life_expectancy/data_cleaning.py [hunk 1]
```diff
@@ -1,90 +1,131 @@
...
+    def clean(self, df_raw: pd.DataFrame) -> pd.DataFrame:
+        df_raw = df_raw.copy()
+        # Split first column into 4 columns
+        df_raw[DataCleaner.DECOMPOSED_COLs] = df_raw[
+            DataCleaner.COMPOSED_COL
+        ].str.split(",", expand=True)
+        df_raw = df_raw.drop(columns=[DataCleaner.COMPOSED_COL])
+
+        # Transform data into long format, filter out missings
+        df_final = pd.melt(df_raw, id_vars=DataCleaner.DECOMPOSED_COLs, var_name="year")
```
Comment (MrCordeiro) [inline life_expectancy/data_cleaning.py]: This bit here (when you use DataCleaner.COMPOSED_COL and DataCleaner.COMPOSED_COL )  is in violation of something called the "Dependency Inversion Principle": in this case, the CSVCleaningStrategy class depends on variables defined in the DataCleaner class, which creates a tight coupling between the two classes.

I.e.: using CSVCleaningStrategy outside a DataCleaner instance would certainly fail.

DECOMPOSED_COLs and COMPOSED_COL should probably be either class or instance variables of the CSVCleaningStrategy class and not DataCleaner

File: life_expectancy/data_cleaning.py [hunk 2]
```diff
@@ -1,90 +1,131 @@
...
+                        )
+                    except ValueError as error:
+                        logging.error("Datatype conversion error %s", error)
+                else:
+                    try:
+                        dataframe[col] = dataframe[col].astype(
+                            dtype=dtype, errors="raise"
+                        )  # type: ignore
+                    except ValueError as error:
+                        logging.error("Datatype conversion error %s", error)
```
Comment (MrCordeiro) [inline life_expectancy/data_cleaning.py:90]: These 2 value errors might cause dirty data to be created. My suggestion is to remove them and allow pandas to raise the ValueError

File: life_expectancy/data_cleaning.py [hunk 3]
```diff
@@ -1,90 +1,131 @@
...
+    COMPOSED_COL = "unit,sex,age,geo\\time"
+    DECOMPOSED_COLs = ["unit", "sex", "age", "region"]
 
-# Define variables cleaning  and filtering data
-COMPOSED_COL = "unit,sex,age,geo\\time"
-DECOMPOSED_COLs = ["unit", "sex", "age", "region"]
+    class NoDataException(Exception):
+        """
+        NoDataException: Exception raised when no data is found for a given region
+        """
```
Comment (MrCordeiro) [inline life_expectancy/data_cleaning.py:106]: I don't see this exception being raised in your code anymore. If it's not being used... perhaps we should remove it?

File: life_expectancy/file_handler.py [hunk 4]
```diff
@@ -1,49 +1,138 @@
...
+class FileLoadingStrategy:
+    """Interface for loading files."""
+
+    def load_data(self, input_file_path: Union[str, Path]) -> pd.DataFrame:
+        """Load the data from a file."""
+        raise NotImplementedError
+
+    def load_data_from_content(self, content: str) -> pd.DataFrame:
+        """Load the data from a string."""
+        raise NotImplementedError
```
Comment (MrCordeiro) [inline life_expectancy/file_handler.py]: **Comment**: You could have used `ABC` and `@abstractmethod` in here as well, just as in the CleaningStrategy ;)

File: life_expectancy/file_handler.py [hunk 5]
```diff
@@ -1,49 +1,138 @@
...
+                    with zip_ref.open(file, "r") as file_ref:
+                        file_content = file_ref.read().decode("utf-8")
+                        return strategy.load_data_from_content(file_content)
+        except FileNotFoundError:
+            logging.error("Input file %s not found.", input_file_path)
+            return pd.DataFrame()
+        except zipfile.BadZipFile:
+            logging.error("Input file %s is not a valid zip file.", input_file_path)
+            return pd.DataFrame()
+        return pd.DataFrame()
```
Comment (MrCordeiro) [inline life_expectancy/file_handler.py]:
```suggestion
def load_data(self, input_file_path: Union[str, Path]) -> pd.DataFrame:
    
    if not input_file_path.exists():
        logging.error("Input file %s not found.", input_file_path)
        return pd.DataFrame()
    
    strategies = {
        ".csv": CSVFileLoadingStrategy,
        ".tsv": JSONFileLoadingStrategy
    }
        
    try:
        with zipfile.ZipFile(Path(input_file_path), "r") as zip_ref:
            file = zip_file.namelist()[0]
            file_ext = Path(file).suffix.lower()
            try:
                strategy = strategies[file_ext]()
            except KeyError as err:
                    logging.error("Unsupported file type inside zip: %s", file_ext)
                    raise err
                
            with zip_ref.open(file, "r") as file_ref:
                    file_content = file_ref.read().decode("utf-8")
                    return strategy.load_data_from_content(file_content)

    except zipfile.BadZipFile:
        logging.error("Input file %s is not a valid zip file.", input_file_path)
        return pd.DataFrame()
    
```

I tried to simplify this function:
- Check if the file exists before trying to unzip it
- Use a dictionary instead of an if-else block to remove one indentation level
- Instead of looking all files inside the zip, just open the first one (because the original code was already exiting the loop after the first file read, the results are the same.

Of course, we can always simplify ever further... heheh

For example, instead of 3 strategies: Json, Csv, and Zip, we could use only 2: CSVFileLoadingStrategy and ZippedJsonLoadingStrategy. Your `load_data` could become simply:

```python
def load_data(self, input_file_path: Union[str, Path]) -> pd.DataFrame:
    """Load a Zipped JSON data and returns a pandas DataFrame."""
        with zipfile.ZipFile(input_file_path) as zip_file:
            zipped_fname = zip_file.namelist()[0]
            with zip_file.open(zipped_fname) as json_file:
                df = pd.read_json(json_file)
        return df

**Review by MrCordeiro — CHANGES_REQUESTED** [review]
If, on one hand the number of tests indeed exploded, on the other hand, one can see that now you master the art of testing! 🥇 

I still get the feeling you are trying to catch all possible exceptions even in cases where they should arguably be raised. This for sure make is harder to write test.

Writing tests is a good litmus code of how easy your code is to read. Take indentations for example, the more indented is the code, the harder it is to follow its logic and edge cases - getting a full test coverage algo gets harder, as each new logical branch is a new test to be written.

(We should have testing activities on smartpairing btw!)

**Review by MrCordeiro — APPROVED** [review]

**Comment by MrCordeiro** [comment]
bravo-applause
